### PR TITLE
Refactor MatTabNav as ARIA tabs with a new MatTabNavPanel component and [tabPanel] input.

### PR DESCRIPTION
**Pull request by zelliott**
Fixes https://github.com/angular/components/issues/23706.

* Adds a new `MatTabNavPanel` component that is intended to be used with `MatTabNav`. `MatTabNavPanel` renders the `tabpanel`, `MatTabNav` renders the `tablist`.
* Adds a new optional `[tabPanel]` input to `MatTabNav` that accepts a `MatTabNavPanel`. Allows `MatTabNav` to make changes to `MatTabNavPanel` (i.e. `aria-labelledby`).
* If `[tabPanel]` is provided, `MatTabNav` acts as an ARIA tabs. Otherwise, it continues to act as links within a navigation landmark (no change from status quo).
* Adds unit tests for new behavior.
* Change made in both non-MDC and MDC tabs.

### CODE REVIEW

File: src/material/tabs/tab-nav-bar/tab-nav-bar.ts [hunk 1]
```diff
@@ -92,6 +95,13 @@ export abstract class _MatTabNavBase
   /** Theme color of the nav bar. */
   @Input() color: ThemePalette = 'primary';
 
+  /**
+   * Associated tab panel controlled by the nav bar. If not provided, then the nav bar
+   * follows the ARIA link / navigation landmark pattern. If provided, it follows the
+   * ARIA tabs design pattern.
+   */
+  @Input() panel?: MatTabNavPanel;
```
Comment (crisbeto) [inline src/material/tabs/tab-nav-bar/tab-nav-bar.ts:103]: The input name might need to be prefixed since the component uses an attribute selector.

File: src/material/tabs/tab-nav-bar/tab-nav-bar.ts [hunk 2]
```diff
@@ -292,12 +335,17 @@ export class _MatTabLinkBase
...
     '[attr.aria-disabled]': 'disabled',
-    '[attr.tabIndex]': 'tabIndex',
+    '[attr.aria-selected]': '_tabNavBar.panel ? (active ? "true" : "false") : null',
+    '[attr.id]': 'id',
+    '[attr.tabIndex]': '_getTabIndex()',
+    '[attr.role]': '_tabNavBar.panel ? "tab" : null',
     '[class.mat-tab-disabled]': 'disabled',
     '[class.mat-tab-label-active]': 'active',
     '(focus)': '_handleFocus()',
+    '(keydown.space)': '_handleSpace()',
```
Comment (crisbeto) [inline src/material/tabs/tab-nav-bar/tab-nav-bar.ts:348]: We don't support any more browsers where `keydown.<key>` doesn't work, but for consistency we should continue using `keyCode` for now. We'll clean all of them up eventually.

File: src/material/tabs/tab-nav-bar/tab-nav-bar.ts [hunk 3]
```diff
@@ -274,7 +293,31 @@ export class _MatTabLinkBase
...
+  _getTabIndex(): number {
+    if (!this._tabNavBar.panel) {
+      return this.tabIndex;
+    }
+
+    if (!this._tabNavBar._items) {
+      return this._isActive ? 0 : -1;
+    }
+
+    return this._tabNavBar.focusIndex === this._getIndex() ? 0 : -1;
```
Comment (crisbeto) [inline src/material/tabs/tab-nav-bar/tab-nav-bar.ts:315]: Do we really need to change the tabindex based on the focus index? If focus is already on the tab, the tabindex doesn't matter anymore.

File: src/material/tabs/tab-nav-bar/tab-nav-bar.ts [hunk 4]
```diff
@@ -324,3 +372,27 @@ export class MatTabLink extends _MatTabLinkBase implements OnDestroy {
...
   }
 }
+
+/**
+ * Tab panel component associated with MatTabNav.
+ */
+@Component({
+  selector: 'mat-tab-nav-panel',
+  exportAs: 'matTabNavPanel',
+  template: '<ng-content select="router-outlet"></ng-content>',
```
Comment (crisbeto) [inline src/material/tabs/tab-nav-bar/tab-nav-bar.ts:382]: I'm not sure that we should be limiting this to `router-outlet`. People could be using a different router implementation.

File: src/material/tabs/tab-nav-bar/tab-nav-bar.ts [hunk 5]
```diff
@@ -324,3 +372,27 @@ export class MatTabLink extends _MatTabLinkBase implements OnDestroy {
...
+  host: {
+    '[attr.aria-labelledby]': '_activeTabId',
+    '[attr.id]': 'id',
+    'role': 'tabpanel',
+  },
+  encapsulation: ViewEncapsulation.None,
+  // tslint:disable-next-line:validate-decorators
+  changeDetection: ChangeDetectionStrategy.Default,
```
Comment (crisbeto) [inline src/material/tabs/tab-nav-bar/tab-nav-bar.ts:390]: Why is this necessary?

File: src/material/tabs/tab-nav-bar/tab-nav-bar.ts (hunk 2 above)
Comment (zelliott) [reply src/material/tabs/tab-nav-bar/tab-nav-bar.ts]: Done.

File: src/material/tabs/tab-nav-bar/tab-nav-bar.ts (hunk 4 above)
Comment (zelliott) [reply src/material/tabs/tab-nav-bar/tab-nav-bar.ts]: Done.

File: src/material/tabs/tab-nav-bar/tab-nav-bar.ts (hunk 5 above)
Comment (zelliott) [reply src/material/tabs/tab-nav-bar/tab-nav-bar.ts]: Changed to `OnPush`.

File: src/material/tabs/tab-nav-bar/tab-nav-bar.ts (hunk 3 above)
Comment (zelliott) [reply src/material/tabs/tab-nav-bar/tab-nav-bar.ts:316]: Currently, `MatTabNav` has by default `tabindex="0"` on all tabs. This means that every tab is in the tab order, which is incorrect for managed focus reasons (per the ARIA spec). Thus, this logic implements the [roving tabindex pattern](https://www.w3.org/TR/wai-aria-practices-1.1/#kbd_roving_tabindex).

Note that `MatTabGroup`'s roving tabindex implementation is actually slightly incorrect, this one I added to `MatTabNav` _is_ correct. It's important for the `tabindex="0"` to track focus, as opposed to selection. This way if you tab to a `MatTabNav`, arrow among some tabs, tab out, and tab back in, the same tab you were previously focused on is still focused.

File: src/material/tabs/tab-nav-bar/tab-nav-bar.ts (hunk 1 above)
Comment (zelliott) [reply src/material/tabs/tab-nav-bar/tab-nav-bar.ts:104]: Do you mind clarifying? Not sure what change you're requesting, sorry.

Comment (crisbeto) [reply src/material/tabs/tab-nav-bar/tab-nav-bar.ts:104]: Sorry, what I meant is that the component currently uses an attribute selector which means that it could have other directives attached to it so the `panel` name could conflict with something else. I was suggesting changing the name to something like `tabPanel`.

File: src/material/tabs/tab-nav-bar/tab-nav-bar.ts (hunk 3 above)
Comment (crisbeto) [reply src/material/tabs/tab-nav-bar/tab-nav-bar.ts:316]: Okay, that makes sense, but I think that `_getIndex` might have to be optimized. Since it's used in a binding, it'll be called on each change detection and it currently uses a linear search to figure out the index. Maybe it can be saved in a property and updated when the list of tabs changes?

Comment (zelliott) [reply src/material/tabs/tab-nav-bar/tab-nav-bar.ts]: Done.

Comment (zelliott) [reply src/material/tabs/tab-nav-bar/tab-nav-bar.ts]: Done.

File: src/material/tabs/tab-nav-bar/tab-nav-bar.ts (hunk 1 above)
Comment (zelliott) [reply src/material/tabs/tab-nav-bar/tab-nav-bar.ts]: Done.

File: src/material/tabs/tab-nav-bar/tab-nav-bar.ts [hunk 6]
```diff
@@ -151,6 +183,7 @@ export abstract class _MatTabNavBase
   templateUrl: 'tab-nav-bar.html',
   styleUrls: ['tab-nav-bar.css'],
   host: {
+    '[attr.role]': 'tabPanel ? "tablist" : null',
```
Comment (zelliott) [inline src/material/tabs/tab-nav-bar/tab-nav-bar.ts:186]: @crisbeto This - as well as some of the `aria-*` attribute host bindings changes below - are breaking changes. For example, if a consumer has added `role="<something>"` to this element, then after this PR is submitted, this host binding will override their specified `role` and set it to `null` (as no `tabPanel` will have been provided). We can avoid this breaking change by modifying this logic to check to see if a `role` has been provided, and if so, use that. For example:

```
'[attr.role]': 'existingRole || (tabPanel ? "tablist" : null)'
```

We'd want to do the same to the other added `aria-*` host bindings. WDYT?

Comment (zelliott) [reply src/material/tabs/tab-nav-bar/tab-nav-bar.ts:186]: Okay.

File: src/material/tabs/tab-nav-bar/tab-nav-bar.ts (hunk 3 above)
Comment (zelliott) [reply src/material/tabs/tab-nav-bar/tab-nav-bar.ts]: I ended up reverting the roving tabindex behavior because it's actually tricky to implement with `FocusKeyManager` without triggering ECAC errors. The problem is that the tabindex needs to be changed on focus, and it's common for people to move focus in `ngAfterViewInit` lifecycles as well as other problematic contexts.

Note that I actually ran into this a year or so ago with `MatSelectionList`. From https://github.com/angular/components/pull/18445:

> I would prefer to use the "roving tabindex" pattern, but given the list options are content-projected, I ran into issues keeping tabindices updated without ExpressionChangedAfter... errors. Thus, I instead essentially followed the implementation used in MatChipList. The listbox element has tabindex="0", and the option elements have tabindex="-1". When the listbox is focused, it redirects focus to the proper option element.

Instead of doing the same thing here, I decided to just match `MatTabGroup`'s behavior and have the tabindex rove on selection as opposed to focus. The reason I decided to do this was because this PR is large as-is. We can always improve the roving tabindex behavior in both tabs components in subsequent PRs.

File: src/material/tabs/tab-nav-bar/tab-nav-bar.ts (hunk 6 above)
Comment (zelliott) [reply src/material/tabs/tab-nav-bar/tab-nav-bar.ts]: Updated the PR per our conversation to fallback to the existing ARIA host attributes if no `tabPanel` is provided, as opposed to removing those attributes with a `null` attribute binding. This makes this PR now truly 100% opt-in & backwards compatible. Consumer code should not change if `tabPanel` isn't provided. This makes it _much_ easier to land internally.

After this PR is merged in, I'll:
1. Migrate the internal instances of consumers overriding ARIA host attributes to the new `tabPanel` input & associated component.
2. Remove the code in this PR that "falls back to the existing ARIA host attributes if no `tabPanel` is provided" (as it calls the DOM directly and not something we want in the component long-term).
3. Migrate the remaining consumer instances onto the new `tabPanel` input & associated component.
4. Remove the code in this PR that falls back to the ARIA link / navigation landmark pattern if no `tabPanel` is provded.

File: src/material-experimental/mdc-tabs/tab-nav-bar/tab-nav-bar.spec.ts [hunk 7]
```diff
@@ -309,6 +315,123 @@ describe('MDC-based MatTabNavBar', () => {
...
 
+  describe('without panel', () => {
+    let fixture: ComponentFixture<SimpleTabNavBarTestApp>;
+
+    beforeEach(() => {
+      fixture = TestBed.createComponent(SimpleTabNavBarTestApp);
+      fixture.detectChanges();
+    });
+
+    it('should have no explicit roles', () => {
```
Comment (crisbeto) [inline src/material-experimental/mdc-tabs/tab-nav-bar/tab-nav-bar.spec.ts:326]: nit: if I were writing these tests, I'd combine all the `with panel` and `without panel` attribute checks into a single test since they're basically checking a single use case.

File: src/material-experimental/mdc-tabs/testing/tab-link-harness.ts [hunk 8]
```diff
@@ -33,6 +33,12 @@ export class MatTabLinkHarness extends ComponentHarness {
     return (await this.host()).text();
   }
 
+  /** Gets the value of the `aria-controls` attribute. */
+  async getAriaControls(): Promise<string | null> {
```
Comment (crisbeto) [inline src/material-experimental/mdc-tabs/testing/tab-link-harness.ts:37]: I don't think that we should be exposing this as a public API. You should be able to do `link.host().getAttribute()` in the nav bar harness.

File: src/material-experimental/mdc-tabs/testing/tab-nav-panel-harness.ts [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {ComponentHarness, HarnessPredicate} from '@angular/cdk/testing';
+import {TabNavPanelHarnessFilters} from './tab-harness-filters';
+
+/** Harness for interacting with a standard mat-tab-nav-panel in tests. */
+export class MatTabNavPanelHarness extends ComponentHarness {
```
Comment (crisbeto) [inline src/material-experimental/mdc-tabs/testing/tab-nav-panel-harness.ts:13]: For components like this one that have projected content, we usually extend the `ContentContainerComponentHarness`.

File: src/material/tabs/tab-nav-bar/tab-nav-bar.ts [hunk 10]
```diff
@@ -238,6 +259,9 @@ export class _MatTabLinkBase
     );
   }
 
+  /** Unique id for the tab. Provided by the sub-class. */
+  id: string;
```
Comment (crisbeto) [inline src/material/tabs/tab-nav-bar/tab-nav-bar.ts:263]: You can set this as `@Input` here in base class so you don't have to override it in the subclasses.

File: src/material/tabs/tab-nav-bar/tab-nav-bar.ts [hunk 11]
```diff
@@ -274,6 +298,42 @@ export class _MatTabLinkBase
...
+    return this._tabNavBar.tabPanel
+      ? this._tabNavBar.tabPanel?.id
+      : this.elementRef.nativeElement.getAttribute('aria-controls');
+  }
+
+  _getAriaSelected(): string | null {
+    if (this._tabNavBar.tabPanel) {
+      return this.active ? 'true' : 'false';
+    } else {
+      return this.elementRef.nativeElement.getAttribute('aria-selected');
```
Comment (crisbeto) [inline src/material/tabs/tab-nav-bar/tab-nav-bar.ts:318]: Instead of calling `getAttribute` for these every time, I think that we can inject them using `@Attribute` in the constructor. I _think_ that our binding will override any pre-existing binding anyway.

Comment (zelliott) [reply src/material/tabs/tab-nav-bar/tab-nav-bar.ts:318]: Originally I tried to use `@Attribute`. The problem is that if someone adds an attribute template binding like `[attr.role]="role"` (as opposed to role="tablist"), then the `@Attribute` binding is actually `null`, because presumably the attribute template binding evaluates too late. That means we can't figure out what `role` they've added without using `getAttribute`. Here's a Stackblitz: https://stackblitz.com/edit/angular-ivy-d2ehto?file=src/app/app.component.ts.

File: src/material-experimental/mdc-tabs/tab-nav-bar/tab-nav-bar.spec.ts (hunk 7 above)
Comment (zelliott) [reply src/material-experimental/mdc-tabs/tab-nav-bar/tab-nav-bar.spec.ts:326]: Acknowledged. I took a stab at refactoring these specs into a two specs (one for with panel and one for without) and they became very large and hard to work with. I'm going to keep them as-is.

File: src/material/tabs/tab-nav-bar/tab-nav-bar.ts (hunk 10 above)
Comment (zelliott) [reply src/material/tabs/tab-nav-bar/tab-nav-bar.ts:263]: Even if I add `@Input` here, I still need to override it in the sub classes because the non-MDC and MDC tabs need a differently formatted unique ID (one of the IDs has `mdc` in it, the other does not). I'm following the pattern of other components, but happy to do things differently.

File: src/material-experimental/mdc-tabs/testing/tab-nav-panel-harness.ts (hunk 9 above)
Comment (zelliott) [reply src/material-experimental/mdc-tabs/testing/tab-nav-panel-harness.ts:13]: Done.

File: src/material-experimental/mdc-tabs/testing/tab-link-harness.ts (hunk 8 above)
Comment (zelliott) [reply src/material-experimental/mdc-tabs/testing/tab-link-harness.ts]: Done.

File: src/material/tabs/tab-nav-bar/tab-nav-bar.ts (hunk 10 above)
Comment (crisbeto) [reply src/material/tabs/tab-nav-bar/tab-nav-bar.ts:263]: The ID pattern doesn't really matter, as long as it's unique. We should be able to avoid ID collisions if we only use the counter in the file with the base class.

File: src/material/tabs/tab-nav-bar/tab-nav-bar.ts (hunk 11 above)
Comment (crisbeto) [reply src/material/tabs/tab-nav-bar/tab-nav-bar.ts:318]: Hmm, interesting. I would've assumed that `getAttribute` will eventually start to return `null` because we clear it in our host bindings.

Comment (zelliott) [reply src/material/tabs/tab-nav-bar/tab-nav-bar.ts:318]: It's not clear to me what you mean here. Can we mark this as resolved and use `getAttribute` or talk more over DM?

File: src/material/tabs/tab-nav-bar/tab-nav-bar.ts (hunk 10 above)
Comment (zelliott) [reply src/material/tabs/tab-nav-bar/tab-nav-bar.ts]: SGTM, updated. Also updated the non-MDC and MDC panels to use the same ID prefix.

File: src/material/tabs/tab-nav-bar/tab-nav-bar.ts (hunk 11 above)
Comment (crisbeto) [reply src/material/tabs/tab-nav-bar/tab-nav-bar.ts:318]: Yeah, let's resolve it. I'm just surprised that it works.

File: src/components-examples/material/tabs/tab-nav-bar-with-panel/tab-nav-bar-with-panel-example.ts [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+  selector: 'tab-nav-bar-with-panel-example',
+  templateUrl: 'tab-nav-bar-with-panel-example.html',
+  styleUrls: ['tab-nav-bar-with-panel-example.css'],
+})
+export class TabNavBarWithPanelExample {
+  links = ['First', 'Second', 'Third'];
+  activeLink = this.links[0];
+  background: ThemePalette = undefined;
+
+  toggleBackground() {
```
Comment (crisbeto) [inline src/components-examples/material/tabs/tab-nav-bar-with-panel/tab-nav-bar-with-panel-example.ts:17]: nit: I think that we can get rid of the background and dynamic links parts of the example since it's only meant to showcase the tab panel.

**Review by crisbeto — APPROVED** [review]
LGTM, the last comment isn't a blocker.

File: src/components-examples/material/tabs/tab-nav-bar-with-panel/tab-nav-bar-with-panel-example.ts (hunk 12 above)
Comment (zelliott) [reply src/components-examples/material/tabs/tab-nav-bar-with-panel/tab-nav-bar-with-panel-example.ts]: Done.

**Review by jelbourn — APPROVED** [review]
Read through the summary and did a quick pass on the code, overall LGTM.
